percona/controllers/pgc/update.go

```diff
@@ -0,0 +1,216 @@
+package pgc
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/robfig/cron/v3"
+	log "github.com/sirupsen/logrus"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/percona/percona-postgresql-operator/internal/kubeapi"
+	"github.com/percona/percona-postgresql-operator/percona/controllers/pgcluster"
+	"github.com/percona/percona-postgresql-operator/percona/controllers/pgreplica"
+	"github.com/percona/percona-postgresql-operator/percona/controllers/version"
+	crv1 "github.com/percona/percona-postgresql-operator/pkg/apis/crunchydata.com/v1"
+)
+
+const never = "never"
+const disabled = "disabled"
+
+func (c *Controller) updateVersion(oldCluster, newCluster *crv1.PerconaPGCluster) error {
+	if oldCluster.Spec.Backup.Image == newCluster.Spec.Backup.Image &&
+		oldCluster.Spec.Backup.BackrestRepoImage == newCluster.Spec.Backup.BackrestRepoImage &&
+		oldCluster.Spec.PGBadger.Image == newCluster.Spec.PGBadger.Image &&
+		oldCluster.Spec.PGBouncer.Image == newCluster.Spec.PGBouncer.Image &&
+		oldCluster.Spec.PGPrimary.Image == newCluster.Spec.PGPrimary.Image {
+		return nil
+	}
+
+	return smartUpdateCluster(c.Client, newCluster, oldCluster)
+}
+
+func (c *Controller) deleteEnsureVersion(jobName string) {
+	c.crons.crons.Remove(cron.EntryID(c.crons.ensureVersionJobs[jobName].ID))
+	delete(c.crons.ensureVersionJobs, jobName)
+}
+
+func (c *Controller) scheduleUpdate(newCluster *crv1.PerconaPGCluster) error {
+	jn := newCluster.Name + "/ensureVersion"
+	schedule, ok := c.crons.ensureVersionJobs[jn]
+	if newCluster.Spec.UpgradeOptions == nil {
+		if ok {
+			c.deleteEnsureVersion(jn)
+		}
+		return nil
+	}
+	if strings.ToLower(newCluster.Spec.UpgradeOptions.Apply) == never ||
+		strings.ToLower(newCluster.Spec.UpgradeOptions.Apply) == disabled {
+		if ok {
+			c.deleteEnsureVersion(jn)
+		}
+	}
+	if ok && schedule.CronSchedule == newCluster.Spec.UpgradeOptions.Schedule {
+		return nil
+	}
+	if ok {
+		c.deleteEnsureVersion(jn)
+	}
+	id, err := c.crons.crons.AddFunc(newCluster.Spec.UpgradeOptions.Schedule, func() {
+		for i := 1; i < 30; i++ {
```

Comment:
In PXC operator we're using mutexes for locking: https://github.com/percona/percona-xtradb-cluster-operator/blob/main/pkg/controller/pxc/version.go#L67-L68

Reply:
https://github.com/percona/percona-postgresql-operator/pull/152/commits/e38bcfc390af0fec4e8bc50743b0a8a38ea841d2